staging/src/k8s.io/apimachinery/pkg/runtime/serializer/cbor/internal/modes/buffers.go

```diff
@@ -0,0 +1,60 @@
+/*
+Copyright 2024 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package modes
+
+import (
+	"bytes"
+	"sync"
+)
+
+var buffers BufferProvider
+
+type buffer struct {
+	bytes.Buffer
+}
+
+type BufferProvider struct {
+	p sync.Pool
+}
+
+func (b *BufferProvider) Get() *buffer {
+	if buf, ok := b.p.Get().(*buffer); ok {
+		return buf
+	}
+	return &buffer{}
+}
+
+func (b *BufferProvider) Put(buf *buffer) {
+	buf.Reset()
+	if buf.Cap() > 3*1024*1024 /* Default MaxRequestBodyBytes */ {
```

Comment:
Curious how to pick a number like this.  I see https://github.com/valyala/bytebufferpool has some benchmarks, but I haven't looked closely enough to assess how useful they are.

Reply:
This number seemed likely to do a good enough job partitioning single objects and small lists from large lists since it's derived from the etcd limits.

I didn't dig too deeply given the number of moving pieces for alpha. It looks like that benchmark in paritcular writes strings of approximately equal length, so it may not be representative of a bimodal pattern where a small fraction of the buffers need to be orders of magnitude larger than the rest.

Reply:
I think I can get behind this as a stop gap for alpha.  It would be great to see this handled upstream in the cbor libs to prevent unbounded memory leakage, but this at least provides us with a bounds.